Lint the less files of the simple theme

return42 opened the pull request:
## What does this PR do?

Lint the less files of the simple theme

About details read along the commit messages.

This PR is limited to the simple theme (I hope we can drop the oscar theme, see #162)

return42 committed:
[mod] add stylelint process to target themes.simple

Before build, do lint the LESS files using stylelint [1].  The configuration
'stylelint-config-standard' [2] is added to the root of the simple theme [3].

[1] https://stylelint.io/
[2] https://github.com/stylelint/stylelint-config-standard
[3] https://stylelint.io/user-guide/configure

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[fix] simple theme: use stylint to fix common lint errors

This fix was autogenerated by::

     npx stylelint -f unix --fix 'searx/static/themes/simple/src/less/**/*.less'

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[stylelint] simple theme: fix some errors reported by stylelint

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[stylelint] disable role 'no-descending-specificity'

This patch disables role 'no-descending-specificity'.  IMO it is better to have
this rule active (see below [1]), but it is hard to rewrite the less files to
pass this rule, so for the first I chose to disable this rule.

---

Source order is important in CSS, and when two selectors have the same
specificity, the one that occurs last will take priority. However, the situation
is different when one of the selectors has a higher specificity. In that case,
source order does not matter: the selector with higher specificity will win out
even if it comes first.

The clashes of these two mechanisms for prioritization, source order and
specificity, can cause some confusion when reading stylesheets. If a selector
with higher specificity comes before the selector it overrides, we have to think
harder to understand it, because it violates the source order
expectation. Stylesheets are most legible when overriding selectors always come
after the selectors they override. That way both mechanisms, source order and
specificity, work together nicely.

This rule enforces that practice as best it can, reporting fewer errors than it
should. It cannot catch every actual overriding selector, but it can catch
certain common mistakes.

[1] https://stylelint.io/user-guide/rules/list/no-descending-specificity/

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

dalf commented on manage:
```diff
@@ -488,8 +488,12 @@ themes.oscar() {
+        build_msg LINT "theme: simple"
+        npm --prefix searx/static/themes/simple run stylelint
+        build_msg GRUNT "theme: simple"
+        npm --prefix searx/static/themes/simple run build
```
`grunt build` already calls `jshint`: 
https://github.com/searxng/searxng/blob/056550d523ddf4e06821957e13623fc322cf4675/searx/static/themes/simple/gruntfile.js#L197

so I would add the call `stylelint` from `gruntfile.js` using [grunt-stylelint](https://www.npmjs.com/package/grunt-stylelint):
```js
grunt.registerTask('default', ['jshint', 'stylelint', 'copy', 'concat', 'uglify', 'less:development', 'less:production']); 
```

return42 replied:
I have seen that there is a grunt-stylelint, but it uses an old ["stylelint": "13.8.0"](https://github.com/wikimedia/grunt-stylelint/commit/5c07ec3767145aeafae043db8d5d3f95f055aec3) while I decided to use the latest versions of stylelint & stylelint-config-standard in our [package.json](https://github.com/searxng/searxng/pull/168/commits/e22dbdda53445008c6b089365c3ba152413a76bc#diff-0f087f6a54013c2b6e21074952c187930cd193591ec780748f18423271435f90).

Do you see a disadvantage in this? ..

Foregive me if I am naive and not used to work with grunt: My impression is that the grunt packages are partly outdated and not updated regularly, so I asking myself why not directly use the npm packages (npm run configuration).  What I mean: we have these shell scripts to build our product, do we realy need additionally grunt packages which wrap old libraries? .. is my thought wrong, what do you think?

dalf replied:
The dependency is ["stylelint": "^13.8.0"](https://github.com/wikimedia/grunt-stylelint/blob/9e56310596ec39c587f5ef20361a73dd0be9fb40/package.json#L26)

See https://docs.npmjs.com/about-semantic-versioning#using-semantic-versioning-to-specify-update-types-your-package-can-accept

`npm install grunt-stylelint stylelint --save-dev` (from https://www.npmjs.com/package/grunt-stylelint), installs the last version:
```js
{
    "grunt-stylelint": "^0.16.0",
    "stylelint": "^13.13.1",
    "stylelint-config-standard": "^22.0.0"
}
```

return42 replied:
> "stylelint": "^13.8.0" .. installs the last version

Argh, you are right .. I have overseen the `^` and thought they are using old versions .. sorry, I am unexercised with npm .. forget my comment, I will patch to use grunt.

return42 replied:
Now I added a stylelint task::

```js
    stylelint: {
      options: {
        formatter: 'unix',
      },
      src: [
        'src/less/**/*.less',
      ]
    },
```

dalf commented on searx/static/themes/simple/package.json:
```diff
@@ -25,6 +27,7 @@
     "webfont": "grunt webfont",
-    "clean": "rm -Rf node_modules package-lock.json ion.less"
+    "clean": "rm -Rf node_modules package-lock.json ion.less",
+    "stylelint": "stylelint -f unix 'src/less/**/*.less'"
```
Based on the comment on `manage` :
```js
  "stylelint": "grunt stylelint",
  "stylelint-fix": "grunt --fix stylelint"
```


https://www.npmjs.com/package/grunt-stylelint#fix

return42 replied:
> `"stylelint-fix": "grunt --fix stylelint"`

It seems `"stylelint-fix": "grunt stylelint --fix"`did also work .. I haven't found a documentation which describes what arguments from the command line are passed through .. what I asking myself is what about common arguments like `--help` or `-v`, these are grunt options and stylint has also the same options.  I looked at:
 
- https://gruntjs.com/using-the-cli and
- https://gruntjs.com/api/grunt.option

dalf commented on searx/static/themes/simple/src/less/style.less:
```diff
@@ -704,7 +704,7 @@ article.result-images[data-vim-selected]::before {
   #linkto_preferences {
     display: none;
-    postion: fixed !important;
+    postion: fixed !important; // stylelint-disable property-no-unknown
```
There is a typo `postion` --> `position` !

return42 replied:
LOL :-) .. here you can see that I am a dyslexic .. I rebased and dropped this commit 50495757, the fix is now in 5b7e7aa.

dalf replied:
Actually I compiled the theme, start the devtools, saw the property was not recognized, and then I realized the problem was not the value but the property name...

dalf reviewed (commented):
Expect the 3 comments LGTM.

I've compiled the theme and look at the result: everything is okay.

return42 committed:
[emacs] .dir-locals.el: add setup for js-mode

- move jshint option from gruntfile to .jshintrc
- remove trailing-whitespace from gruntfile and
- add jshint esversion: 6
- .dir-locals.el add locals for js-mode to use JSHint from the simple theme

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>
